src/compiler.rkt

```diff
@@ -276,356 +276,289 @@
-    [(member op (list ival-sin ival-cos ival-sinh ival-cosh))
-     ; log[Гcos] = log[x] + log[sin(x)] - log[cos(x)] <= log[x] - log[cos(x)]
-     ; log[Гsin] = log[x] + log[cos(x)] - log[sin(x)] <= log[x] - log[sin(x)]
-     ;                      ^^^^^^^^^^^                         ^^^^^^^^^^^^^
-     ;                      can be pruned                       a possible uncertainty
+    [(ival-sin ival-cos ival-sinh ival-cosh)
+     ; log[Гcos] = log[x] + log[sin(x)] - log[cos(x)] <= log[x] - log[cos(x)] + 1
+     ; log[Гsin] = log[x] + log[cos(x)] - log[sin(x)] <= log[x] - log[sin(x)] + 1
+     ;                                                          ^^^^^^^^^^^^^
+     ;                                                          a possible uncertainty
+     (define x (first srcs))
      (define outlo (ival-lo output))
      (define outhi (ival-hi output))
 
-     (define out-exp
-       (+ (min (log2-approx outlo)
-               (log2-approx outhi))
-          (if (equal? (bigfloat-signbit outlo) (bigfloat-signbit outhi))
-              0                         ; both bounds are positive or negative
-              (- (get-slack)))))        ; Condition of uncertainty,
-     ; slack is negated because it is to be subtracted below
-     (max 0 (+ 1 ; subtraction of logarithms doesn't consider mantissa - which can be +1 to the result
-               (- (max
-                   (log2-approx (ival-lo (car srcs)))
-                   (log2-approx (ival-hi (car srcs))))
-                  out-exp)))]
+     (define x-exp (ival-max-log2-approx x))          ; log[x]
+     (define out-exp (min (log2-approx outlo)         ; log[cos(x)] or log[sin(x)]
+                          (log2-approx outhi)))
+     
+     (define slack (if (equal? (mpfr-sign outlo) (mpfr-sign outhi))
+                       0
+                       (get-slack)))                  ; Condition of uncertainty
+     
+     (list (max 0 (+ (- x-exp out-exp) 1 slack)))]
                
-    [(member op (list ival-log ival-log2 ival-log10))
+    [(ival-log ival-log2 ival-log10)
      ; log[Гlog]   = log[1/logx] = -log[log(x)]
      ; log[Гlog2]  = log[1/(log2(x) * ln(2))] <= -log[log2(x)] + 1    < main formula
      ; log[Гlog10] = log[1/(log10(x) * ln(10))] <= -log[log10(x)] - 1
-     ;                    ^ a possible uncertainty
+     ; log[Гlog10] < log[Гlog] < log[Гlog2]
+     
      (define outlo (ival-lo output))
      (define outhi (ival-hi output))
-       
-     (max 0
-          (+ 1 (max (- (log2-approx outlo))  ; main formula
-                    (- (log2-approx outhi)))
-             (if (equal? (bigfloat-signbit outlo) (bigfloat-signbit outhi)) ; slack part
-                 0                           ; both bounds are positive or negative
-                 (get-slack))))]             ; output crosses 0.bf - uncertainty
-
-          
+     (define out-exp (max (- (log2-approx outlo))     ; -log[log10(x)]
+                          (- (log2-approx outhi))))
```

Comment:
I think this is `(- (ival-min-log2-approx output))`?

Reply:
Ahh, these max mins are tricky!